Script for Jenkins parameterized matrix build

caisq opened the pull request:
The script takes a few environment variables as input and invoke
ci_build.sh in the same directory in the proper way. It takes
information about the build type (cpu | gpu | android), Python version,
whether "-c opt" is used and whether PIP install-test is to be
performed. It automatically examines the parameters to determine the
correct build script to call and the argument flags to use with it. It
also examines whether Docker is available on the system to determine
whether the build will be done inside Docker or not. If GPU Docker build
is detectd, it'll map the CUDA devices and libraries automatically.
In addition, it looks at some optional environment variables to set
additional command flags for Docker and the build command.

See experimental Jenkins matrix builds at: 
For CPU (Linux and Mac): http://ci.tensorflow.org/view/Experimental/job/experimental-cais-matrix-linux-cpu-test/8/
For GPU (Linux): http://ci.tensorflow.org/view/Experimental/job/experimental-cais-matrix-linux-gpu-test/
For Android: http://ci.tensorflow.org/view/Experimental/job/experimental-cais-matrix-linux-android-test/

caisq committed:
Script for Jenkins parameterized matrix build

The script takes a few environment variables as input and invoke
ci_build.sh in the same directory in the proper way. It takes
information about the build type (cpu | gpu | android), Python version,
whether "-c opt" is used and whether PIP install-test is to be
performed. It automatically examines the parameters to determine the
correct build script to call and the argument flags to use with it. It
also examines whether Docker is available on the system to determine
whether the build will be done inside Docker or not. If GPU Docker build
is detectd, it'll map the CUDA devices and libraries automatically.
In addition, it looks at some optional environment variables to set
additional command flags for Docker and the build command.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/tools/ci_build/ci_parameterized_build.sh:
```diff
@@ -0,0 +1,204 @@
+#
+# Note: certain combinations of parameter values are regarded
+# as invalid and will cause the script to exit with code 0. For example:
+#   NO_OPT & PIP     (PIP builds should always use OPT)
```
Why is that?

caisq replied:
See my answer to @jendap below.

martinwicke commented:
I like this. @jendap?

jendap commented on tensorflow/tools/ci_build/ci_parameterized_build.sh:
```diff
@@ -0,0 +1,204 @@
+NO_DOCKER_MAIN_CMD="${CI_BUILD_DIR}/builds/configured"
+
+# Additional option flags to apply when Docker is unavailable (e.g., on Mac)
+NO_DOCKER_OPT_FLAG="-j 1 --genrule_strategy=standalone "\
```
Do we need these arguments outside of mac? Why just one job? It will take forever to build, will it not? And why the other two?

martinwicke replied:
If it works, let's omit the -j1.

On Thu, Feb 4, 2016 at 4:33 PM Jan Prach <notifications@github.com> wrote:

> In tensorflow/tools/ci_build/ci_parameterized_build.sh
> <https://github.com/tensorflow/tensorflow/pull/998#discussion_r51961149>:
>
> > +str_strip () {
> > +  echo -e "$1" | sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//'
> > +}
> > +
> > +
> > +##########################################################
> > +# Default configuration
> > +CI_BUILD_DIR="tensorflow/tools/ci_build"
> > +
> > +# Command to call when Docker is available
> > +DOCKER_MAIN_CMD="${CI_BUILD_DIR}/ci_build.sh"
> > +# Command to call when Docker is unavailable
> > +NO_DOCKER_MAIN_CMD="${CI_BUILD_DIR}/builds/configured"
> > +
> > +# Additional option flags to apply when Docker is unavailable (e.g., on Mac)
> > +NO_DOCKER_OPT_FLAG="-j 1 --genrule_strategy=standalone "\
>
> Do we need these arguments outside of mac? Why just one job? It will take
> forever to build, will it not? And why the other two?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/998/files#r51961149>.
>

caisq replied:
I copied those from the mac build on Jenkins assuming they were necessary. I confirmed that except "--linkopt=-headerpad_max_install_names", the flags can be omitted without causing the build to fail. The "--linkopt" flag seems to be necessary. See: 
http://ci.tensorflow.org/job/tensorflow-master-mac/146/

Keeping it ensures that both the clean and non-clean bazel build pass.
http://ci.tensorflow.org/job/tensorflow-master-mac/147/
http://ci.tensorflow.org/job/tensorflow-master-mac/148/

So I'll remove the option flags except "--linkopt=-headerpad_max_install_names".

martinwicke replied:
I've had trouble without --spawn_strategy=standalone but a lot has changed
since then.
On Thu, Feb 4, 2016 at 19:14 caisq <notifications@github.com> wrote:

> In tensorflow/tools/ci_build/ci_parameterized_build.sh
> <https://github.com/tensorflow/tensorflow/pull/998#discussion_r51972755>:
>
> > +str_strip () {
> > +  echo -e "$1" | sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//'
> > +}
> > +
> > +
> > +##########################################################
> > +# Default configuration
> > +CI_BUILD_DIR="tensorflow/tools/ci_build"
> > +
> > +# Command to call when Docker is available
> > +DOCKER_MAIN_CMD="${CI_BUILD_DIR}/ci_build.sh"
> > +# Command to call when Docker is unavailable
> > +NO_DOCKER_MAIN_CMD="${CI_BUILD_DIR}/builds/configured"
> > +
> > +# Additional option flags to apply when Docker is unavailable (e.g., on Mac)
> > +NO_DOCKER_OPT_FLAG="-j 1 --genrule_strategy=standalone "\
>
> I copied those from the mac build on Jenkins assuming they were necessary.
> I confirmed that except "--linkopt=-headerpad_max_install_names", the flags
> can be omitted without causing the build to fail. The "--linkopt" flag
> seems to be necessary. See:
> http://ci.tensorflow.org/job/tensorflow-master-mac/146/
>
> Keeping it ensures that both the clean and non-clean bazel build pass.
> http://ci.tensorflow.org/job/tensorflow-master-mac/147/
> http://ci.tensorflow.org/job/tensorflow-master-mac/148/
>
> So I'll remove the option flags except
> "--linkopt=-headerpad_max_install_names".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/998/files#r51972755>.
>

jendap commented on tensorflow/tools/ci_build/ci_parameterized_build.sh:
```diff
@@ -0,0 +1,204 @@
+  OPT_FLAG="${OPT_FLAG} ${NO_DOCKER_OPT_FLAG}"
+
+elif [[ ${CTYPE} == "gpu" ]]; then  # Docker GPU: Map devices and libraries
+  CUDA_DEVICES=$(ls /dev/nvidia* | xargs -I{} echo '--device {}:{}' | xargs)
```
You do not have to do this. It is already in ci_build.sh

caisq replied:
Removed as suggested.

jendap commented on tensorflow/tools/ci_build/ci_parameterized_build.sh:
```diff
@@ -0,0 +1,204 @@
+if [[ ${TF_BUILD_IS_OPT} == "no_opt" ]]; then
+  # PIP builds are done only with the -c opt flag
+  if [[ ${TF_BUILD_IS_PIP} == "pip" ]]; then
+    echo "Skipping parameter combination: ${TF_BUILD_IS_OPT} & "\
```
Why not?

caisq replied:
When you generate the install package, typically you want to use the optimized C++ build. If there is any build configuration that we want to leave out for efficiency, these are the first ones to go.

caisq replied:
In fact, if you look at pip.sh (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/builds/pip.sh), the "-c opt" option is always used when running the target "//tensorflow/tools/pip_package:build_pip_package". So specifying NO_OPT here would have no effect anyway.

caisq commented:
I addressed comments from @jendap. In addition, I realized that there was an issue with BUILD_TAG values from matrix builds. Those tags have special characters such as "," and "=", which make the Docker image names illegal. I modified ci_build.sh a little to solve that problem.

caisq commented:
@martinwicke Please hold on merging this one. There are some issues related to Docker runs for parameterized builds. I'm looking into it. I'll let you know when it's okay to merge.

martinwicke commented:
Ok
On Thu, Feb 4, 2016 at 20:29 caisq <notifications@github.com> wrote:

> @martinwicke <https://github.com/martinwicke> Please hold on merging this
> one. There are some issues related to Docker runs for parameterized builds.
> I'm looking into it. I'll let you know when it's okay to merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/998#issuecomment-180188710>
> .
>

caisq commented on tensorflow/tools/ci_build/ci_build.sh:
```diff
@@ -35,7 +35,9 @@ fi
-CI_DOCKER_EXTRA_PARAMS=("${CI_DOCKER_EXTRA_PARAMS[@]:---rm}")
+if [[ "${CI_DOCKER_EXTRA_PARAMS}" != *"--rm"* ]]; then
+  CI_DOCKER_EXTRA_PARAMS="--rm ${CI_DOCKER_EXTRA_PARAMS}"
+fi
```
Making sure that the "--rm" flag for Docker is always set. In the case of automatic parametric build, CI_DOCKER_EXTRA_PARAMS may already been set to an non-empty value. So the previous approach needs to be replaced.

caisq commented:
@martinwicke @jendap Feel free to merge now. The issue on Jenkins matrix build is now resolved. The change set is now squashed and ready.

caisq committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into param-build-2

caisq committed:
Adding option for parallel build and serial tests

Potentially useful for limiting the build time for GPU builds, where the
tests cannot run in parallel due to contention or GPU memory.

caisq committed:
Fixing issues and debugging

caisq committed:
Fixing more issues

caisq committed:
Fixing more

caisq committed:
Fixing more 2

martinwicke commented:
I merged this, apparently five hours ago.
